### ISSUE TITLE
Floating Point Constants

### ISSUE THREAD

**Issue by damianmoz** [chapel-lang/chapel#18599]
Various `generic programming issues` (which is largely unobtrusive currently) will arise as Chapel has to adapt to handling either real(16) or real(128) found in some of the latest CPUs. Indeed, one only has to look at what you have to do in C/C++ when trying to generically program to handle floating point constants for float, double and long double.

The only way around this is to assume that Chapel effectively defines a floating constant as having infinite precision until it is used opr at least the maximum precision handled by the hardware, or some _compiler-flag-restricted_ maximum precision . The latter still allows us to mandate that a floating point constant is 64-bits wide so existing programs are not broken. The programmer can still specify the size occupied by that constant from the start of execution by a cast as in
```chapel
const SQRT_2_0 = 1.41421356237309504880168872420969807856967187537694807317667973799:real(w);
```
The question then arises of how you specify the precision of a symbolic constant. These are things like **INFINITY**, **e** and **pi**(both of which drive me crazy because they conflict with names my programs use), and others.

This also will mean that the compiler will need a flag which allows a programmer to specify the rounding mode used at compile time to
an expression like the above or the one below:
```chapel
param t = 1 / 3.0:real(32);
```
Similarly, is the expression on the right-hand side of this,
```chapel
const t = 1 / 3.0:real(32);
```
evaluated at compile time and determined by that previously mentioned flag or by the rounding flag in effect at run-time,. or is it determined by a separate compiler flag?

Let us see where this discussion goes. I am sure this discussion has implications for the optimiser.

This discssion may lead to the need  to define a parameterised type with say
```chapel
type zuse(w) = uint(w);
```
in Chapel. But it may not.

**Comment by damianmoz** [comment]
Sorry, I was a bit brief in my previous reply. And I am sure that the following is not the last discussion we will need. 

I am sure there WILL be lots of potential AI users of Chapel who need **real**(16). Me personally at the moment - No. When I start using Chapel on hardware with 128-bit floating point at some FUTURE date, I personally WILL have a need for more precision than **real**(64) on a regular basis to validate algorithms that I wrote with **real**(64). To date, I achieve this now, I compromise by writing my algorithms in **real**(32) and then validating with **real**(64). Or, I revert to developing the algorithm in C and doing everything with **double** and then validating with **long double**. And then I port to Chapel. Double work. Messy. But there will come a time when your user-base WILL have hardware without 80-bit floating point numbers and who WILL want to develop with 64-bit floating point and then WILL want to validate with 128-bit floating point.

And if Chapel is lumbered with design decisions which have made the incorporation of **binary16** and **binary128** floating point awkward, it is shooting itself in the foot and either will loose users or not gain them.

On another point, we need, probably sooner rather than later, a compiler option to flag expressions with a mix of types, either mixed mode or implicit type conversion. Given a generic type _R_, consider:
```chapel
const x = 20:R;
const y = x + 1.0/3.0;
const third = 1.0/3.0;
const a = x + third;
param third_ = third;
const a_ = x + third_;
```
If R is **real**(32), the types of **a** and **a_** are different. That does not name sense to me. Depending on what I am doing, one of those is wrong and finding such an error is hard. I probably wanted the type of `a` to be `R`, just like `a_`. To achieve this, I need to do
```chapel
const third = (1.0/3.0):R; // rounding done as per prevailing rounding mode
const a = x + third;
```
That way, I am guaranteed that the calculation will be done in **real**(32) arithmetic - which is what I want. No implicit conversions. In the earlier calculation of `a_`, I am assuming it is done in **real**(64) arithmetic which is not what I want.

Note that for completeness, I need to write:
```chapel
param third_ = (1.0/3.0):R; // rounding done as per compiler option, default (could be) round to nearest, ties to even
const a_ = x + third_; // type of a is R and computation of a_ is done at the precision of R
```

**Comment by damianmoz** [comment]
That issue with **INFINITY** or **Infinity** or **infinity** or **inf** or **Inf** or whatever it gets called in Chapel 2.0, needs to get a type change. It should be the smallest type supported by the underlying hardware modulo the Chapel compiler.

In the same timeframe and before things get cast in stone, there should be some thought given to supporting a **qNaN** and an **sNaN**. Exactly how is best. But C23 is a good starting point.

Also, out of curiousity, does the idea of running Chapel on a machine that does not support an Infinity and the various **NaN**s make any sense? To me it does not but what my choice is by definition selfish.

**Comment by mppf** [comment]
>  It should be the smallest type supported by the underlying hardware modulo the Chapel compiler.

Is the type of `inf` particularly important to you? And if so, why is `real(32)` more intuitive than `real(64)`?

`INFINITY` is currently a param, so it should be able to implicitly convert to `real(32)` just as `1.0` would. Note that `1.0` is a param with type `real(64)`. The current situation of `INFINITY` being a param with type `real(64)` makes it consistent with a literal like `1.0`.

I think this current situation is *better* for a future when we add `real(16)` and `real(128)` vs. changing the type of `INFINITY` to depend on the compilation target.

**Comment by damianmoz** [comment]
In C, **INFINITY** and **NAN** were declared as 32-bits which meant that when they appeared in a 32-bit expression, it did not automatically promote the expression to a higher precision, for example to 64-bits.

It seems the same occurs in Chapel 1.28.0 built with LLVM 14.0.0, Brad;;s special version which supports the new primitive for #14140 . If I do
```chapel
const two = 2:real(32);
const t = INFINITY * two;
```
then the **type** of _t_ is **real**(64) because it is the mix of two items, a **real**(32) and a **real**(64).

Whereas, if the type of **INFINITY** was the smallest supported by the hardware (and Chapel) which in the future could even be **real**(16), then it would be dominated by the type of the variable _two_ and hence the type of _t_ would, in the above case be what I desired which is **real**(32).

Or have I losr the plot entirely?

**Comment by mppf** [comment]
Oh, *that*. You are correct that the program you show does not behave the way you (or we) want it to. But it's just an oversight/bug rather than something wrong with the approach.

Here is a more complete reproducer:

``` chapel
const two = 2:real(32);
const t = INFINITY * two;

writeln("2:real(32) two = ", two, " : ", two.type:string);
writeln("INFINITY*two t = ", t, " : ", t.type:string);
```

PR #20995 fixes the oversight to get it printing out real(32) for the type of both numbers.


> then the **type** of _t_ is **real**(64) because it is the mix of two items, a **real**(32) and a **real**(64).

Right but as we have discussed elsewhere, for an operation that is a mix of a literal (or params) and a variable (e.g. `var x: real(32); var y = x * 12.3;`), it is desireable to ignore the literal/param type in favor of the variable's type. Since `INFINITY` is a param here, its type should be ignored.

**Comment by damianmoz** [comment]
You are right. I thought we had discussed it elsewhere but I could not find it and wondered whether my memory was playing tricks on me. Also, I was not searching for the word `narrowing` either which did not help.

**Comment by bradcray** [comment]
Catching up on this issue, and having resolved the bug about infinity's ability to narros in #20995, I'd like to hide the comments that are now resolved to simplify the issue for those coming to it later.  But I'm not certain how far back is safe to hide.  I think https://github.com/chapel-lang/chapel/issues/18599#issuecomment-1305766704 seems safe to hide?  But maybe the previous comment could be as well?

**Comment by damianmoz** [comment]
In hindsight, if I knew then what was in the pipeline what I know now, this issue would be very, very tiny.
